File: src/connectivity.py
"""
Connectivity analyzer for KiCad schematics.

Builds a net list by tracing wire connections between component pins.
"""

from dataclasses import dataclass, field
from collections import defaultdict
from typing import Optional

from kicad_sch import Schematic, Component, Wire, Junction, Label, Point


@dataclass
class WireSegment:
    """A wire segment with its points."""
    points: list[tuple[float, float]]  # List of (x, y) coordinates


@dataclass
class Net:
    """A net connecting multiple pins."""
    name: str
    pins: list[tuple[str, str, str]] = field(default_factory=list)  # (ref, pin_number, pin_name)
    wires: list[WireSegment] = field(default_factory=list)  # Wire segments in this net
    is_power: bool = False


@dataclass
class Netlist:
    """Complete netlist extracted from a schematic."""
    nets: list[Net] = field(default_factory=list)
    components: list[Component] = field(default_factory=list)


def analyze_connectivity(sch: Schematic, tolerance: float = 0.01) -> Netlist:
    """
    Analyze schematic connectivity and build a netlist.

    Algorithm:
    1. Build a graph of all wire segments and junction points
    2. Find connected components in this graph (each is a potential net)
    3. For each connected component, find all pins that touch it
    4. Assign net names from labels or generate anonymous names
    """
    # Build point-to-segment mapping
    # Each point can connect to multiple wire segments
    point_to_segments: dict[Point, list[int]] = defaultdict(list)

    for i, wire in enumerate(sch.wires):
        for pt in wire.points:
            point_to_segments[pt].append(i)

    # Add junctions as explicit connection points
    junction_points = {Point(j.x, j.y) for j in sch.junctions}

    # Build adjacency: which segments connect to which
    segment_adj: dict[int, set[int]] = defaultdict(set)

    for pt, segments in point_to_segments.items():
        # All segments sharing this point are connected
        for i in segments:
            for j in segments:
                if i != j:
                    segment_adj[i].add(j)

    # Also connect segments that share endpoints (wires can chain)
    for i, wire_i in enumerate(sch.wires):
        for j, wire_j in enumerate(sch.wires):
            if i >= j:
                continue
            # Check if any endpoint of wire_i matches any endpoint of wire_j
            for pt_i in wire_i.points:
                for pt_j in wire_j.points:
                    if points_match(pt_i, pt_j, tolerance):
                        segment_adj[i].add(j)
                        segment_adj[j].add(i)

    # Find connected components of wire segments using union-find
    parent = list(range(len(sch.wires)))

    def find(x):
        if parent[x] != x:
            parent[x] = find(parent[x])
        return parent[x]

    def union(x, y):
        px, py = find(x), find(y)
        if px != py:
            parent[px] = py

    for i, neighbors in segment_adj.items():
        for j in neighbors:
            union(i, j)

    # Group segments by their root
    segment_groups: dict[int, list[int]] = defaultdict(list)
    for i in range(len(sch.wires)):
        segment_groups[find(i)].append(i)

    # For each group, collect all points
    group_points: dict[int, set[Point]] = {}
    for root, segments in segment_groups.items():
        pts = set()
        for seg_idx in segments:
            for pt in sch.wires[seg_idx].points:
                pts.add(pt)
        group_points[root] = pts

    # Build pin position to (ref, pin_number, pin_name) mapping
    pin_positions: dict[Point, list[tuple[str, str, str]]] = defaultdict(list)
    for comp in sch.components:
        # Look up pin names from lib_symbol
        lib_sym = sch.lib_symbols.get(comp.lib_id)
        pin_names = {}
        if lib_sym:
            for pin in lib_sym.pins:
                pin_names[pin.number] = pin.name

        for pin_num, pin_pos in comp.pins.items():
            pin_name = pin_names.get(pin_num, '')
            pin_positions[pin_pos].append((comp.reference, pin_num, pin_name))

    # Build label position to label name mapping
    label_positions: dict[Point, str] = {}
    for label in sch.labels:
        label_positions[Point(label.x, label.y)] = label.name

    # For each wire group, find connected pins and labels
    nets: list[Net] = []
    anonymous_counter = 1

    for root, points in group_points.items():
        pins: list[tuple[str, str, str]] = []
        net_name: Optional[str] = None
        is_power = False

        for pt in points:
            # Check for pins at this point (with tolerance)
            for pin_pt, pin_list in pin_positions.items():
                if points_match(pt, pin_pt, tolerance):
                    pins.extend(pin_list)

            # Check for labels at this point
            for label_pt, label_name in label_positions.items():
                if points_match(pt, label_pt, tolerance):
                    net_name = label_name

        # Check if any connected pin is a power symbol
        for ref, pin_num, pin_name in pins:
            comp = next((c for c in sch.components if c.reference == ref), None)
            if comp and comp.lib_id in sch.lib_symbols:
                if sch.lib_symbols[comp.lib_id].is_power:
                    # Use the value as the net name (e.g., "+5V", "GND")
                    net_name = comp.value
                    is_power = True
                    break

        # Remove power symbol pins from the pin list (they define nets, not connections)
        pins = [(ref, pin_num, pin_name) for ref, pin_num, pin_name in pins
                if not (any(c.reference == ref and c.lib_id in sch.lib_symbols and
                          sch.lib_symbols[c.lib_id].is_power
                          for c in sch.components))]

        # Skip empty nets
        if not pins:
            continue

        # Assign anonymous name if needed
        if net_name is None:
            net_name = f"N{anonymous_counter}"
            anonymous_counter += 1

        # Remove duplicates while preserving order
        seen = set()
        unique_pins = []
        for pin in pins:
            if pin not in seen:
                seen.add(pin)
                unique_pins.append(pin)

        # Collect wire segments for this net
        wire_segments = []
        for seg_idx in segment_groups[root]:
            wire = sch.wires[seg_idx]
            segment_points = [(p.x, p.y) for p in wire.points]
            wire_segments.append(WireSegment(points=segment_points))

        nets.append(Net(name=net_name, pins=unique_pins, wires=wire_segments, is_power=is_power))

    # Filter out components that are power symbols
    real_components = [c for c in sch.components
                       if not (c.lib_id in sch.lib_symbols and
                              sch.lib_symbols[c.lib_id].is_power)]

    # Merge nets with the same name
    merged_nets: dict[str, Net] = {}
    for net in nets:
        if net.name in merged_nets:
            # Merge pins and wires into existing net
            existing = merged_nets[net.name]
            for pin in net.pins:
                if pin not in existing.pins:
                    existing.pins.append(pin)
            existing.wires.extend(net.wires)
            existing.is_power = existing.is_power or net.is_power
        else:
            merged_nets[net.name] = net
    nets = list(merged_nets.values())

    # Sort nets: power first, then named, then anonymous
    def net_sort_key(net: Net):
        name = net.name
        if net.is_power:
            # Power nets first, sorted by voltage (descending for positive, ascending for negative)
            if name.startswith('+'):
                try:
                    voltage = float(''.join(c for c in name if c.isdigit() or c == '.'))
                    return (0, -voltage, name)
                except:
                    return (0, 0, name)
            elif name == 'GND' or name.startswith('GND'):
                return (1, 0, name)
            elif name.startswith('-'):
                try:
                    voltage = float(''.join(c for c in name if c.isdigit() or c == '.'))
                    return (2, voltage, name)
                except:
                    return (2, 0, name)
            else:
                return (0, 0, name)
        elif name.startswith('N') and name[1:].isdigit():
            # Anonymous nets last
            return (4, int(name[1:]), name)
        else:
            # Named nets in between
            return (3, 0, name)

    nets.sort(key=net_sort_key)

    # Sort pins within each net
    for net in nets:
        net.pins.sort(key=lambda p: (p[0], int(p[1]) if p[1].isdigit() else 0, p[2]))

    return Netlist(nets=nets, components=real_components)


def points_match(p1: Point, p2: Point, tolerance: float = 0.01) -> bool:
    """Check if two points are within tolerance of each other."""
    return abs(p1.x - p2.x) < tolerance and abs(p1.y - p2.y) < tolerance


def print_netlist(netlist: Netlist):
    """Print the netlist in a readable format."""
    print("Components:")
    for comp in netlist.components:
        print(f"  {comp.reference}: {comp.lib_id} = {comp.value}")

    print("\nNets:")
    for net in netlist.nets:
        pins_str = ", ".join(f"{ref}.{pin_num}:{pin_name}" if pin_name else f"{ref}.{pin_num}"
                             for ref, pin_num, pin_name in net.pins)
        power_marker = " [POWER]" if net.is_power else ""
        print(f"  {net.name}: {pins_str}{power_marker}")


if __name__ == '__main__':
    import sys
    from kicad_sch import parse_schematic

    if len(sys.argv) < 2:
        print("Usage: python connectivity.py <schematic.kicad_sch>")
        sys.exit(1)

    sch = parse_schematic(sys.argv[1])
    print(f"Analyzing: {sch.title}\n")

    netlist = analyze_connectivity(sch)
    print_netlist(netlist)

File: src/ts/connectivity.ts
/**
 * Connectivity analyzer for KiCad schematics.
 *
 * Builds a net list by tracing wire connections between component pins.
 */

import {
  Schematic,
  Component,
  Point,
  pointHash,
} from './kicadSch';

export interface WireSegment {
  points: [number, number][]; // List of [x, y] coordinates
}

export interface Net {
  name: string;
  pins: [string, string, string][]; // [ref, pin_number, pin_name]
  wires: WireSegment[];
  isPower: boolean;
}

export interface Netlist {
  nets: Net[];
  components: Component[];
}

/**
 * Check if two points are within tolerance of each other.
 */
function pointsMatch(p1: Point, p2: Point, tolerance: number = 0.01): boolean {
  return Math.abs(p1.x - p2.x) < tolerance && Math.abs(p1.y - p2.y) < tolerance;
}

/**
 * Analyze schematic connectivity and build a netlist.
 *
 * Algorithm:
 * 1. Build a graph of all wire segments and junction points
 * 2. Find connected components in this graph (each is a potential net)
 * 3. For each connected component, find all pins that touch it
 * 4. Assign net names from labels or generate anonymous names
 */
export function analyzeConnectivity(sch: Schematic, tolerance: number = 0.01): Netlist {
  // Build point-to-segment mapping
  // Each point can connect to multiple wire segments
  const pointToSegments = new Map<string, number[]>();

  for (let i = 0; i < sch.wires.length; i++) {
    for (const pt of sch.wires[i].points) {
      const key = pointHash(pt);
      if (!pointToSegments.has(key)) {
        pointToSegments.set(key, []);
      }
      pointToSegments.get(key)!.push(i);
    }
  }

  // Build adjacency: which segments connect to which
  const segmentAdj = new Map<number, Set<number>>();

  for (const [, segments] of pointToSegments) {
    // All segments sharing this point are connected
    for (const i of segments) {
      for (const j of segments) {
        if (i !== j) {
          if (!segmentAdj.has(i)) {
            segmentAdj.set(i, new Set());
          }
          segmentAdj.get(i)!.add(j);
        }
      }
    }
  }

  // Also connect segments that share endpoints (wires can chain)
  for (let i = 0; i < sch.wires.length; i++) {
    for (let j = i + 1; j < sch.wires.length; j++) {
      // Check if any endpoint of wire_i matches any endpoint of wire_j
      for (const ptI of sch.wires[i].points) {
        for (const ptJ of sch.wires[j].points) {
          if (pointsMatch(ptI, ptJ, tolerance)) {
            if (!segmentAdj.has(i)) {
              segmentAdj.set(i, new Set());
            }
            if (!segmentAdj.has(j)) {
              segmentAdj.set(j, new Set());
            }
            segmentAdj.get(i)!.add(j);
            segmentAdj.get(j)!.add(i);
          }
        }
      }
    }
  }

  // Find connected components of wire segments using union-find
  const parent = Array.from({ length: sch.wires.length }, (_, i) => i);

  function find(x: number): number {
    if (parent[x] !== x) {
      parent[x] = find(parent[x]);
    }
    return parent[x];
  }

  function union(x: number, y: number): void {
    const px = find(x);
    const py = find(y);
    if (px !== py) {
      parent[px] = py;
    }
  }

  for (const [i, neighbors] of segmentAdj) {
    for (const j of neighbors) {
      union(i, j);
    }
  }

  // Group segments by their root
  const segmentGroups = new Map<number, number[]>();
  for (let i = 0; i < sch.wires.length; i++) {
    const root = find(i);
    if (!segmentGroups.has(root)) {
      segmentGroups.set(root, []);
    }
    segmentGroups.get(root)!.push(i);
  }

  // For each group, collect all points
  const groupPoints = new Map<number, Set<string>>();
  const groupPointsList = new Map<number, Point[]>();
  for (const [root, segments] of segmentGroups) {
    const pts = new Set<string>();
    const ptsList: Point[] = [];
    for (const segIdx of segments) {
      for (const pt of sch.wires[segIdx].points) {
        const key = pointHash(pt);
        if (!pts.has(key)) {
          pts.add(key);
          ptsList.push(pt);
        }
      }
    }
    groupPoints.set(root, pts);
    groupPointsList.set(root, ptsList);
  }

  // Build pin position to (ref, pin_number, pin_name) mapping
  const pinPositions = new Map<string, [string, string, string][]>();
  for (const comp of sch.components) {
    // Look up pin names from lib_symbol
    const libSym = sch.libSymbols.get(comp.libId);
    const pinNames = new Map<string, string>();
    if (libSym) {
      for (const pin of libSym.pins) {
        pinNames.set(pin.number, pin.name);
      }
    }

    for (const [pinNum, pinPos] of comp.pins) {
      const pinName = pinNames.get(pinNum) || '';
      const key = pointHash(pinPos);
      if (!pinPositions.has(key)) {
        pinPositions.set(key, []);
      }
      pinPositions.get(key)!.push([comp.reference, pinNum, pinName]);
    }
  }

  // Build label position to label name mapping
  const labelPositions = new Map<string, string>();
  for (const label of sch.labels) {
    labelPositions.set(pointHash({ x: label.x, y: label.y }), label.name);
  }

  // For each wire group, find connected pins and labels
  const nets: Net[] = [];
  let anonymousCounter = 1;

  for (const [root, points] of groupPointsList) {
    const pins: [string, string, string][] = [];
    let netName: string | null = null;
    let isPower = false;

    for (const pt of points) {
      // Check for pins at this point (with tolerance)
      for (const [pinKey, pinList] of pinPositions) {
        const [px, py] = pinKey.split(',').map(parseFloat);
        if (pointsMatch(pt, { x: px, y: py }, tolerance)) {
          pins.push(...pinList);
        }
      }

      // Check for labels at this point
      for (const [labelKey, labelName] of labelPositions) {
        const [lx, ly] = labelKey.split(',').map(parseFloat);
        if (pointsMatch(pt, { x: lx, y: ly }, tolerance)) {
          netName = labelName;
        }
      }
    }

    // Check if any connected pin is a power symbol
    for (const [ref, ,] of pins) {
      const comp = sch.components.find(c => c.reference === ref);
      if (comp) {
        const libSym = sch.libSymbols.get(comp.libId);
        if (libSym?.isPower) {
          // Use the value as the net name (e.g., "+5V", "GND")
          netName = comp.value;
          isPower = true;
          break;
        }
      }
    }

    // Remove power symbol pins from the pin list (they define nets, not connections)
    const filteredPins = pins.filter(([ref, ,]) => {
      const comp = sch.components.find(c => c.reference === ref);
      if (comp) {
        const libSym = sch.libSymbols.get(comp.libId);
        return !libSym?.isPower;
      }
      return true;
    });

    // Skip empty nets
    if (filteredPins.length === 0) {
      continue;
    }

    // Assign anonymous name if needed
    if (netName === null) {
      netName = `N${anonymousCounter}`;
      anonymousCounter++;
    }

    // Remove duplicates while preserving order
    const seen = new Set<string>();
    const uniquePins: [string, string, string][] = [];
    for (const pin of filteredPins) {
      const key = `${pin[0]}.${pin[1]}`;
      if (!seen.has(key)) {
        seen.add(key);
        uniquePins.push(pin);
      }
    }

    // Collect wire segments for this net
    const wireSegments: WireSegment[] = [];
    const segments = segmentGroups.get(root) || [];
    for (const segIdx of segments) {
      const wire = sch.wires[segIdx];
      const segmentPoints: [number, number][] = wire.points.map(p => [p.x, p.y]);
      wireSegments.push({ points: segmentPoints });
    }

    nets.push({ name: netName, pins: uniquePins, wires: wireSegments, isPower });
  }

  // Filter out components that are power symbols
  const realComponents = sch.components.filter(c => {
    const libSym = sch.libSymbols.get(c.libId);
    return !libSym?.isPower;
  });

  // Merge nets with the same name
  const mergedNets = new Map<string, Net>();
  for (const net of nets) {
    if (mergedNets.has(net.name)) {
      // Merge pins and wires into existing net
      const existing = mergedNets.get(net.name)!;
      for (const pin of net.pins) {
        const key = `${pin[0]}.${pin[1]}`;
        const existingKeys = new Set(existing.pins.map(p => `${p[0]}.${p[1]}`));
        if (!existingKeys.has(key)) {
          existing.pins.push(pin);
        }
      }
      existing.wires.push(...net.wires);
      existing.isPower = existing.isPower || net.isPower;
    } else {
      mergedNets.set(net.name, net);
    }
  }

  const sortedNets = Array.from(mergedNets.values());

  // Sort nets: power first, then named, then anonymous
  sortedNets.sort((a, b) => {
    const keyA = netSortKey(a);
    const keyB = netSortKey(b);

    for (let i = 0; i < Math.min(keyA.length, keyB.length); i++) {
      if (keyA[i] < keyB[i]) return -1;
      if (keyA[i] > keyB[i]) return 1;
    }
    return 0;
  });

  // Sort pins within each net
  for (const net of sortedNets) {
    net.pins.sort((a, b) => {
      if (a[0] < b[0]) return -1;
      if (a[0] > b[0]) return 1;
      const numA = /^\d+$/.test(a[1]) ? parseInt(a[1], 10) : 0;
      const numB = /^\d+$/.test(b[1]) ? parseInt(b[1], 10) : 0;
      if (numA < numB) return -1;
      if (numA > numB) return 1;
      if (a[2] < b[2]) return -1;
      if (a[2] > b[2]) return 1;
      return 0;
    });
  }

  return { nets: sortedNets, components: realComponents };
}

/**
 * Generate sort key for a net.
 */
function netSortKey(net: Net): (number | string)[] {
  const name = net.name;

  if (net.isPower) {
    if (name.startsWith('+')) {
      const match = name.match(/[\d.]+/);
      const voltage = match ? parseFloat(match[0]) : 0;
      return [0, -voltage, name];
    } else if (name === 'GND' || name.startsWith('GND')) {
      return [1, 0, name];
    } else if (name.startsWith('-')) {
      const match = name.match(/[\d.]+/);
      const voltage = match ? parseFloat(match[0]) : 0;
      return [2, voltage, name];
    } else {
      return [0, 0, name];
    }
  } else if (/^N\d+$/.test(name)) {
    // Anonymous nets last
    return [4, parseInt(name.slice(1), 10), name];
  } else {
    // Named nets in between
    return [3, 0, name];
  }
}
